{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Count codons\n",
    "This notebook takes FASTQ files and uses dms_tools2 to calculate the frequency of codons at mutagenezed sites in the Zika Virus NS5 protein.\n",
    "\n",
    "FASTQ files for each cell population are specified in `samplesheet.csv`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notebook setup\n",
    "Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dms_tools2==2.6.8\n",
      "numpy==1.19.1\n",
      "pandas==0.25.3\n",
      "plotnine==0.6.0\n",
      "regex==2.5.83\n"
     ]
    }
   ],
   "source": [
    "import dms_tools2\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import plotnine as p9\n",
    "import regex\n",
    "from Bio import SeqIO\n",
    "from dms_tools2.ipython_utils import showPDF\n",
    "\n",
    "# print imports and versions\n",
    "# copied from https://stackoverflow.com/questions/40428931/package-for-listing-version-of-packages-used-in-a-jupyter-notebook\n",
    "print('\\n'.join(f'{m.__name__}=={m.__version__}' \\\n",
    "                for m in globals().values() if getattr(m, '__version__', None)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequencing_dir = '/shared/ngs/illumina/bloom_lab/211223_M03100_0703_000000000-K5NB5/Unaligned/Project_bloom_lab/'\n",
    "samplesheet = 'data/samplesheet.csv'\n",
    "reference_file = 'data/ns5.fasta'\n",
    "alignspecs_file = 'data/alignspecs.txt'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Output data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_dir = 'results/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analysis parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notebook aesthetics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "p9.theme_set(p9.theme_classic())\n",
    "CBPALETTE_RICH = ['#648FFF', '#FFB000', '#DC267F', '#785EF0', '#FE6100']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data\n",
    "Load samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>library</th>\n",
       "      <th>source</th>\n",
       "      <th>stage</th>\n",
       "      <th>sorted</th>\n",
       "      <th>r1_file</th>\n",
       "      <th>r1_file_long</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>wt-plasmid-noStage-notSorted</td>\n",
       "      <td>wt</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>noStage</td>\n",
       "      <td>notSorted</td>\n",
       "      <td>BR_WT_Plasmid_S1_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>wt-gDNA-noStage-notSorted</td>\n",
       "      <td>wt</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>noStage</td>\n",
       "      <td>notSorted</td>\n",
       "      <td>BR_WT_NS5_gDNA_S17_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lib1-plasmid-noStage-notSorted</td>\n",
       "      <td>lib1</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>noStage</td>\n",
       "      <td>notSorted</td>\n",
       "      <td>BR_Lib_1_Plasmid_S2_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lib2-plasmid-noStage-notSorted</td>\n",
       "      <td>lib2</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>noStage</td>\n",
       "      <td>notSorted</td>\n",
       "      <td>BR_Lib_2_Plasmid_S3_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lib3-plasmid-noStage-notSorted</td>\n",
       "      <td>lib3</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>noStage</td>\n",
       "      <td>notSorted</td>\n",
       "      <td>BR_Lib_3_Plasmid_S4_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>lib1-gDNA-early-notSorted</td>\n",
       "      <td>lib1</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>early</td>\n",
       "      <td>notSorted</td>\n",
       "      <td>BR_Lib_1_Early_gDNA_S5_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>lib2-gDNA-early-notSorted</td>\n",
       "      <td>lib2</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>early</td>\n",
       "      <td>notSorted</td>\n",
       "      <td>BR_Lib_2_Early_gDNA_S6_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>lib3-gDNA-early-notSorted</td>\n",
       "      <td>lib3</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>early</td>\n",
       "      <td>notSorted</td>\n",
       "      <td>BR_Lib_3_Early_gDNA_S7_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>lib1-gDNA-mid-notSorted</td>\n",
       "      <td>lib1</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>mid</td>\n",
       "      <td>notSorted</td>\n",
       "      <td>BR_Lib_1_Mid_gDNA_S8_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>lib2-gDNA-mid-notSorted</td>\n",
       "      <td>lib2</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>mid</td>\n",
       "      <td>notSorted</td>\n",
       "      <td>BR_Lib_2_Mid_gDNA_S9_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>lib3-gDNA-mid-notSorted</td>\n",
       "      <td>lib3</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>mid</td>\n",
       "      <td>notSorted</td>\n",
       "      <td>BR_Lib_3_Mid_gDNA_S10_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>lib1-gDNA-late-notSorted</td>\n",
       "      <td>lib1</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>late</td>\n",
       "      <td>notSorted</td>\n",
       "      <td>BR_Lib_1_Late_gDNA_S11_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>lib2-gDNA-late-notSorted</td>\n",
       "      <td>lib2</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>late</td>\n",
       "      <td>notSorted</td>\n",
       "      <td>BR_Lib_2_Late_gDNA_S12_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>lib3-gDNA-late-notSorted</td>\n",
       "      <td>lib3</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>late</td>\n",
       "      <td>notSorted</td>\n",
       "      <td>BR_Lib_3_Late_gDNA_S13_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>lib1-gDNA-late-sorted</td>\n",
       "      <td>lib1</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>late</td>\n",
       "      <td>sorted</td>\n",
       "      <td>BR_Lib_1_Late_Sort_gDNA_S14_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>lib2-gDNA-late-sorted</td>\n",
       "      <td>lib2</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>late</td>\n",
       "      <td>sorted</td>\n",
       "      <td>BR_Lib_2_Late_Sort_gDNA_S15_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>lib3-gDNA-late-sorted</td>\n",
       "      <td>lib3</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>late</td>\n",
       "      <td>sorted</td>\n",
       "      <td>BR_Lib_3_Late_Sort_gDNA_S16_R1_001.fastq.gz</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              name library   source    stage     sorted  \\\n",
       "0     wt-plasmid-noStage-notSorted      wt  plasmid  noStage  notSorted   \n",
       "1        wt-gDNA-noStage-notSorted      wt     gDNA  noStage  notSorted   \n",
       "2   lib1-plasmid-noStage-notSorted    lib1  plasmid  noStage  notSorted   \n",
       "3   lib2-plasmid-noStage-notSorted    lib2  plasmid  noStage  notSorted   \n",
       "4   lib3-plasmid-noStage-notSorted    lib3  plasmid  noStage  notSorted   \n",
       "5        lib1-gDNA-early-notSorted    lib1     gDNA    early  notSorted   \n",
       "6        lib2-gDNA-early-notSorted    lib2     gDNA    early  notSorted   \n",
       "7        lib3-gDNA-early-notSorted    lib3     gDNA    early  notSorted   \n",
       "8          lib1-gDNA-mid-notSorted    lib1     gDNA      mid  notSorted   \n",
       "9          lib2-gDNA-mid-notSorted    lib2     gDNA      mid  notSorted   \n",
       "10         lib3-gDNA-mid-notSorted    lib3     gDNA      mid  notSorted   \n",
       "11        lib1-gDNA-late-notSorted    lib1     gDNA     late  notSorted   \n",
       "12        lib2-gDNA-late-notSorted    lib2     gDNA     late  notSorted   \n",
       "13        lib3-gDNA-late-notSorted    lib3     gDNA     late  notSorted   \n",
       "14           lib1-gDNA-late-sorted    lib1     gDNA     late     sorted   \n",
       "15           lib2-gDNA-late-sorted    lib2     gDNA     late     sorted   \n",
       "16           lib3-gDNA-late-sorted    lib3     gDNA     late     sorted   \n",
       "\n",
       "                                        r1_file  \\\n",
       "0              BR_WT_Plasmid_S1_R1_001.fastq.gz   \n",
       "1            BR_WT_NS5_gDNA_S17_R1_001.fastq.gz   \n",
       "2           BR_Lib_1_Plasmid_S2_R1_001.fastq.gz   \n",
       "3           BR_Lib_2_Plasmid_S3_R1_001.fastq.gz   \n",
       "4           BR_Lib_3_Plasmid_S4_R1_001.fastq.gz   \n",
       "5        BR_Lib_1_Early_gDNA_S5_R1_001.fastq.gz   \n",
       "6        BR_Lib_2_Early_gDNA_S6_R1_001.fastq.gz   \n",
       "7        BR_Lib_3_Early_gDNA_S7_R1_001.fastq.gz   \n",
       "8          BR_Lib_1_Mid_gDNA_S8_R1_001.fastq.gz   \n",
       "9          BR_Lib_2_Mid_gDNA_S9_R1_001.fastq.gz   \n",
       "10        BR_Lib_3_Mid_gDNA_S10_R1_001.fastq.gz   \n",
       "11       BR_Lib_1_Late_gDNA_S11_R1_001.fastq.gz   \n",
       "12       BR_Lib_2_Late_gDNA_S12_R1_001.fastq.gz   \n",
       "13       BR_Lib_3_Late_gDNA_S13_R1_001.fastq.gz   \n",
       "14  BR_Lib_1_Late_Sort_gDNA_S14_R1_001.fastq.gz   \n",
       "15  BR_Lib_2_Late_Sort_gDNA_S15_R1_001.fastq.gz   \n",
       "16  BR_Lib_3_Late_Sort_gDNA_S16_R1_001.fastq.gz   \n",
       "\n",
       "                                         r1_file_long  \n",
       "0   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "1   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "2   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "3   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "4   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "5   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "6   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "7   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "8   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "9   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "10  /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "11  /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "12  /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "13  /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "14  /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "15  /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "16  /shared/ngs/illumina/bloom_lab/211223_M03100_0...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "samples = pd.read_csv(samplesheet)\n",
    "samples.insert(0, 'name', (samples['library'] + '-' +\n",
    "                           samples['source'] + '-' +\n",
    "                           samples['stage'] + '-' +\n",
    "                           samples['sorted']))\n",
    "samples['r1_file_long'] = sequencing_dir + samples['r1_file']\n",
    "display(samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load reference sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The reference sequence name is: NS5\n",
      "The nucleotide sequence is:\n",
      "GGAGGTGGGACGGGAGAGACTCTGGGAGAGAAGTGGAAAGCTCGTCTGAATCAGATGTCGGCCCTGGAGTTCTACTCTTATAAAAAGTCAGGTATCACTGAAGTGTGTAGAGAGGAGGCTCGCCGTGCCCTCAAGGATGGAGTGGCCACAGGAGGACATGCCGTATCCCGGGGAAGTGCAAAGCTCAGATGGTTGGTGGAGAGAGGATATCTGCAGCCCTATGGGAAGGTTGTTGACCTCGGATGTGGCAGAGGGGGCTGGAGCTATTATGCCGCCACCATCCGCAAAGTGCAGGAGGTGAGAGGATACACAAAGGGAGGTCCCGGTCATGAAGAACCCATGCTGGTGCAAAGCTATGGGTGGAACATAGTTCGTCTCAAGAGTGGAGTGGACGTCTTCCACATGGCGGCTGAGCCGTGTGACACTCTGCTGTGTGACATAGGTGAGTCATCATCTAGTCCTGAAGTGGAAGAGACACGAACACTCAGAGTGCTCTCTATGGTGGGGGACTGGCTTGAAAAAAGACCAGGGGCCTTCTGTATAAAGGTGCTGTGCCCATACACCAGCACTATGATGGAAACCATGGAGCGACTGCAACGTAGGCATGGGGGAGGATTAGTCAGAGTGCCATTGTCTCGCAACTCCACACATGAGATGTACTGGGTCTCTGGGGCAAAGAGCAACATCATAAAAAGTGTGTCCACCACAAGTCAGCTCCTCCTGGGACGCATGGATGGCCCCAGGAGGCCAGTGAAATATGAGGAGGATGTGAACCTCGGCTCGGGTACACGAGCTGTGGCAAGCTGTGCTGAGGCTCCTAACATGAAAATCATCGGCAGGCGCATTGAGAGAATCCGCAATGAACATGCAGAAACATGGTTTCTTGATGAAAACCACCCATACAGGACATGGGCCTACCATGGGAGCTACGAAGCCCCCACGCAAGGATCAGCGTCTTCCCTCGTGAACGGGGTTGTTAGACTCCTGTCAAAGCCTTGGGACGTGGTGACTGGAGTTACAGGAATAGCCATGACTGACACCACACCATACGGCCAACAAAGAGTCTTCAAAGAAAAAGTGGACACCAGGGTGCCAGATCCCCAAGAAGGCACTCGCCAGGTAATGAACATAGTCTCTTCCTGGCTGTGGAAGGAGCTGGGGAAACGCAAGCGGCCACGCGTCTGCACCAAAGAAGAGTTTATCAACAAGGTGCGCAGCAATGCAGCACTGGGAGCAATATTTGAAGAGGAAAAAGAATGGAAGACGGCTGTGGAAGCTGTGAATGATCCAAGGTTTTGGGCCCTAGTGGATAGGGAGAGAGAACACCACCTGAGAGGAGAGTGTCACAGCTGTGTGTACAACATGATGGGAAAAAGAGAAAAGAAGCAAGGAGAGTTCGGGAAAGCAAAAGGTAGCCGCGCCATCTGGTACATGTGGTTGGGAGCCAGATTCTTGGAGTTTGAAGCTCTTGGATTCTTGAACGAGGACCATTGGATGGGAAGAGAAAACTCAGGAGGTGGAGTCGAAGGGTTAGGATTGCAAAGACTTGGATACATTCTAGAAGAAATGAATCGGGCACCAGGAGGAAAGATGTACGCAGATGACACTGCTGGCTGGGACACCCGCATTAGTAAGTTTGATCTGGAGAATGAAGCTCTGATTACCAACCAAATGGAGGAAGGGCACAGAACTCTGGCGTTGGCCGTGATTAAATACACATACCAAAACAAAGTGGTGAAGGTTCTCAGACCAGCTGAAGGAGGAAAAACAGTTATGGACATCATTTCAAGACAAGACCAGAGAGGGAGTGGACAAGTTGTCACTTATGCTCTCAACACATTCACCAACTTGGTGGTGCAGCTTATCCGGAACATGGAAGCTGAGGAAGTGTTAGAGATGCAAGACTTATGGTTGTTGAGGAAGCCAGAGAAAGTGACCAGATGGTTGCAGAGCAATGGATGGGATAGACTCAAACGAATGGCGGTCAGTGGAGATGACTGCGTTGTGAAGCCAATCGATGATAGGTTTGCACATGCCCTCAGGTTCTTGAATGACATGGGAAAAGTTAGGAAAGACACACAGGAGTGGAAACCCTCGACTGGATGGAGCAATTGGGAAGAAGTCCCGTTCTGCTCCCACCACTTCAACAAGCTGTACCTCAAGGATGGGAGATCCATTGTGGTCCCTTGCCGCCACCAAGATGAACTGATTGGCCGAGCTCGCGTCTCACCAGGGGCAGGATGGAGCATCCGGGAGACTGCCTGTCTTGCAAAATCATATGCGCAGATGTGGCAGCTCCTTTATTTCCACAGAAGAGACCTTCGACTGATGGCTAATGCCATTTGCTCGGCTGTGCCAGTTGACTGGGTACCAACTGGGAGAACCACCTGGTCAATCCATGGAAAGGGAGAATGGATGACCACTGAGGACATGCTCATGGTGTGGAATAGAGTGTGGATTGAGGAGAACGACCATATGGAGGACAAGACTCCTGTAACAAAATGGACAGACATTCCCTATCTAGGAAAAAGGGAGGACTTATGGTGTGGATCCCTTATAGGGCACAGACCCCGCACCACTTGGGCTGAAAACATCAAAGACACAGTCAACATGGTGCGCAGGATCATAGGTGATGAAGAAAAGTACATGGACTATCTATCCACCCAAGTCCGCTACTTGGGTGAGGAAGGGTCCACACCCGGAGTGTTGTAACTCGAGCAGCTGTAGACTTAGCCACTTTTTAAAAGAAAAGGGGGGACTGGAAGGGCTAATTCACTCCCAACGAAGACAAGATATCCTTGATCTGTGGATCTACCACACACAAGGCTACTTCCCTGATTGGCAGAACTACACACCAGGACCAGGGATCAGATATCCACTGACCTTTGGATGGCGCTACAAGCTAGTACCAGTTGAGCCAGAGAAGTTAGAAGAAGCCAACAAAGGAGAGAACACCAGCTTGTTACACCCTGTGAGCCTGCATGG\n"
     ]
    }
   ],
   "source": [
    "reference = SeqIO.read(reference_file, 'fasta')\n",
    "ref_seq = reference.seq\n",
    "\n",
    "print(f'The reference sequence name is: {reference.id}')\n",
    "print(f'The nucleotide sequence is:\\n{ref_seq}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in alignment specs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The alignment specs are:\n",
      "\n",
      "1951,2285,33,32\n"
     ]
    }
   ],
   "source": [
    "with open (alignspecs_file, 'r') as file:\n",
    "    alignspecs = file.read().replace('\\n','')\n",
    "\n",
    "print('The alignment specs are:\\n')\n",
    "print(alignspecs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Count codon frequencies\n",
    "Use `dms_tools2` to count codon frequencies in sequencing data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Setup:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>R1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>wt-plasmid-noStage-notSorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>wt-gDNA-noStage-notSorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lib1-plasmid-noStage-notSorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lib2-plasmid-noStage-notSorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lib3-plasmid-noStage-notSorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>lib1-gDNA-early-notSorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>lib2-gDNA-early-notSorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>lib3-gDNA-early-notSorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>lib1-gDNA-mid-notSorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>lib2-gDNA-mid-notSorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>lib3-gDNA-mid-notSorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>lib1-gDNA-late-notSorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>lib2-gDNA-late-notSorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>lib3-gDNA-late-notSorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>lib1-gDNA-late-sorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>lib2-gDNA-late-sorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>lib3-gDNA-late-sorted</td>\n",
       "      <td>/shared/ngs/illumina/bloom_lab/211223_M03100_0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              name  \\\n",
       "0     wt-plasmid-noStage-notSorted   \n",
       "1        wt-gDNA-noStage-notSorted   \n",
       "2   lib1-plasmid-noStage-notSorted   \n",
       "3   lib2-plasmid-noStage-notSorted   \n",
       "4   lib3-plasmid-noStage-notSorted   \n",
       "5        lib1-gDNA-early-notSorted   \n",
       "6        lib2-gDNA-early-notSorted   \n",
       "7        lib3-gDNA-early-notSorted   \n",
       "8          lib1-gDNA-mid-notSorted   \n",
       "9          lib2-gDNA-mid-notSorted   \n",
       "10         lib3-gDNA-mid-notSorted   \n",
       "11        lib1-gDNA-late-notSorted   \n",
       "12        lib2-gDNA-late-notSorted   \n",
       "13        lib3-gDNA-late-notSorted   \n",
       "14           lib1-gDNA-late-sorted   \n",
       "15           lib2-gDNA-late-sorted   \n",
       "16           lib3-gDNA-late-sorted   \n",
       "\n",
       "                                                   R1  \n",
       "0   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "1   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "2   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "3   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "4   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "5   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "6   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "7   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "8   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "9   /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "10  /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "11  /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "12  /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "13  /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "14  /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "15  /shared/ngs/illumina/bloom_lab/211223_M03100_0...  \n",
       "16  /shared/ngs/illumina/bloom_lab/211223_M03100_0...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing batchfile to results/codon_countsbatchfile.csv\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "# Set up results dir\n",
    "codon_counts_dir = os.path.join(results_dir + 'codon_counts')\n",
    "os.makedirs(codon_counts_dir, exist_ok=True)\n",
    "\n",
    "# Make samplesheet\n",
    "batchfile = os.path.join(codon_counts_dir + 'batchfile.csv')\n",
    "batchfile_list = []\n",
    "for index, sample in samples.iterrows():\n",
    "    batchfile_list.append(sample[['name', 'r1_file_long']])\n",
    "batchfile_df = pd.DataFrame(batchfile_list)\n",
    "batchfile_df = batchfile_df.rename(columns={'r1_file_long': 'R1'})\n",
    "display(batchfile_df)\n",
    "\n",
    "print(f'Writing batchfile to {batchfile}')\n",
    "batchfile_df.to_csv(batchfile, index=False)\n",
    "print('Done.')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running dms_tools2\n",
      "Processed sequencing data to create codon counts files in results/codon_counts\n"
     ]
    }
   ],
   "source": [
    "ncpus = 16\n",
    "\n",
    "R1_trim = 200\n",
    "R2_trim = 200\n",
    "\n",
    "# Overwrite or use existing results\n",
    "use_existing = 'no'\n",
    "\n",
    "print('Running dms_tools2')\n",
    "log = ! dms2_batch_bcsubamp \\\n",
    "        --batchfile {batchfile} \\\n",
    "        --refseq {reference_file} \\\n",
    "        --alignspecs {alignspecs} \\\n",
    "        --outdir {codon_counts_dir} \\\n",
    "        --summaryprefix summary \\\n",
    "        --R1trim {R1_trim} \\\n",
    "        --R2trim {R2_trim} \\\n",
    "        --ncpus {ncpus} \\\n",
    "        --use_existing {use_existing}\n",
    "\n",
    "samples['codon_counts_file'] = codon_counts_dir + '/' + samples['name'] + '_codoncounts.csv'\n",
    "\n",
    "# check that expected codon counts files created\n",
    "assert all(map(os.path.isfile, samples.codon_counts_file)), '\\n'.join(log)\n",
    "\n",
    "print(f\"Processed sequencing data to create codon counts files in {codon_counts_dir}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
